### PR TITLE
Rework the PHP compactor

### LINKED ISSUE #263: Refactor the PHP annotation based compactor
| Question    | Answer
| ------------| ---------------
| Box version | 3.0.0-beta.4
| PHP version | 7.2.6
| Platform with version | Ubuntu 16.04
| Installation type | Composer

Some Composer dependencies (`herrera-io/annotations`) are abandoned. You should use some alternatives.

```
Package herrera-io/annotations is abandoned, you should avoid using it. No replacement was suggested.
```

### PR DESCRIPTION
**Pull request by theofidry**
Closes #263

Goal: remove the need of the deprecated dependency [herrera-io/annotations](https://github.com/kherge-abandoned/php-annotations) and the underlying [doctrine/lexer 1.x](https://packagist.org/packages/doctrine/lexer) which is bound to disappear at some point.

- Remove the deprecated dependency  [herrera-io/annotations](https://github.com/kherge-abandoned/php-annotations) and soon deprecated underlying dependency [doctrine/lexer 1.x](https://packagist.org/packages/doctrine/lexer) in favour of [HoaCompiler](https://hoa-project.net/En/Literature/Hack/Compiler.html) and an annotation grammar which should land in doctrine/annotations or lexer at some point.
- Changes to the `Php` compactor:
    - Invalid annotations are no longer recognised as annotations:
        ```php
        /**
          * @Annotation ()
          * @Namespaced\ Annotation
          */
        ```
        
        Will be transformed into:
        
        ```php
        /**
          * @Annotation
          * @Namespaced
          */
        ```
- The removal of common annotations is enabled by default
- The setting `annotation#ignore` no longer accepts a `string` value, only `string[]` and `null` are allowed
- Upon some annotation parsing failures, the error is thrown to the user in order to identify and fix those cases instead of always silently ignore the error.
- Add recommendations and warnings when relevant
- Update the documentation

### CODE REVIEW

File: res/annotation-grammar.pp [hunk 1]
```diff
@@ -0,0 +1,86 @@
+%pragma lexer.unicode 1
+
+%skip   space              [\x20\x09\x0a\x0d]+
+%token  doc_               /\*\*                      -> docblock
+
+%skip   docblock:space     [\x20\x09\x0a\x0d]+
+%skip   docblock:star      \*(?!/)
+%token  docblock:_doc      \*/                         -> default
+%token  docblock:at        @(?!\s)                     -> annot
```
Comment (Hywan) [inline res/annotation-grammar.pp]: Maybe update the token to this:

```
%token docblock:at (?<!\b)@(?=\b)
```

Testing:

```
(1)  * @bcdef
(2)  * @ bcdef
(3)  * @author <hello@world>
(4)  *@xyz
```

Case 1, the `@` is selected, as expected.
Case 2, the `@` is _not_ selected, as expected.
Case 3, the `@` in `@author` is selected, and the `@` in the email address is _not_ selected, as expected.
Case 4, the `@` is selected, as expected.

See regexr.com/42ktg

Thoughts?

File: src/Annotation/DocblockParser.php [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    public function parse(string $docblock): TreeNode
+    {
+        $docblock = trim($docblock);
+
+        if (0 !== strpos($docblock, '/**') || '*/' !== substr($docblock, -2)) {
+            return new TreeNode('#null');
+        }
+
+        $docblock = preg_replace(
+            '/(\/\*\*[\s\S]*?\@author.+?)(\<.+?\@.+?\>)([\s\S]*?\*\/)/',
```
Comment (Hywan) [inline src/Annotation/DocblockParser.php]: Assuming my previous comment is working, this hack be can removed.

File: res/annotation-grammar.pp (hunk 1 above)
Comment (Hywan) [reply res/annotation-grammar.pp]: I just though that `hello-@world` is valid email address too, so the token might be updated to:

```diff
- %token docblock:at (?<!\b)@(?=\b)
+ %token docblock:at (?<=[\h\*])@(?=\b)
```

Use the same https://regexr.com/42ktg link, with a new test case:

```
(5)  *@xyz <hello-@world>
```

Case 5, the `@` in `@xyz` is selected, and the `@` in the email address is _not_ selected, as expected.

Thoughts?

**Comment by theofidry** [comment]
It's in! Many thanks for the help :)

**Comment by Hywan** [comment]
What was the problem with https://github.com/humbug/box/pull/315/commits/73e853c17b8df17bf19c8ce1be94d49b288de5a5? :-) I just want to understand so that I can improve myself.

**Comment by theofidry** [comment]
Too many cases failing :(

I think the idea is good but it requires further tweaking with the other rules afterwards. I don't remember the specific errors on top of my head. That said, the patch is not lost as @majkl is aware of it and it will still be needed for Doctrine annotations

**Comment by Hywan** [comment]
Thanks :-)!

**Pull request merged**